File: plugins/agento11y/internal/agents/cursor/tags/tags.go
package tags

// BuiltinInputs carries the values used to populate built-in tag keys.
//
// `git.branch` resolution lives in the shared package
// plugins/agento11y/internal/gitbranch so codex/copilot can reuse it without
// importing this cursor-scoped package.
type BuiltinInputs struct {
	WorkspaceRoot     string
	Cwd               string
	Entrypoint        string
	GitBranch         string
	IsBackgroundAgent bool
}

// Build returns the per-generation built-in tags. SIGIL_TAGS-supplied values
// are layered in by the SDK at the client level; built-ins take precedence
// because the SDK merges per-generation Tags atop client-level Tags.
// Returns nil when no built-ins are present.
func Build(in BuiltinInputs) map[string]string {
	out := make(map[string]string, 4)
	if in.GitBranch != "" {
		out["git.branch"] = in.GitBranch
	}
	cwd := in.Cwd
	if cwd == "" {
		cwd = in.WorkspaceRoot
	}
	if cwd != "" {
		out["cwd"] = cwd
	}
	if in.Entrypoint != "" {
		out["entrypoint"] = in.Entrypoint
	}
	if in.IsBackgroundAgent {
		out["subagent"] = "true"
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

File: plugins/pi/src/tags.ts
// Build per-generation built-in tags for the pi plugin.
//
// Mirrors the Go side in plugins/agento11y/internal/agents/cursor/tags/tags.go
// `Build`: only emits keys whose values are non-empty, returns `undefined`
// when no inputs are populated. User-supplied `SIGIL_TAGS` is layered in by
// the SDK at the client level; the seed/built-in tags merged here take
// precedence on collisions because the SDK merges per-generation Tags atop
// client-level Tags.

export interface BuiltinTagInputs {
  cwd?: string;
  gitBranch?: string;
}

export function buildBuiltinTags(
  in_: BuiltinTagInputs,
): Record<string, string> | undefined {
  const out: Record<string, string> = {};
  if (in_.gitBranch) {
    out["git.branch"] = in_.gitBranch;
  }
  if (in_.cwd) {
    out.cwd = in_.cwd;
  }
  if (Object.keys(out).length === 0) {
    return undefined;
  }
  return out;
}
